packages/dds/tree/src/forest/cursor.ts

```diff
@@ -4,89 +4,203 @@
+    /**
+     * Moves `offset` nodes in the field.
+     * If seeking to exactly past either end,
+     * returns false and navigates up to the parent field (setting `inFields` to `true`).
+     *
+     * If mode is `Fields`, offset is from the corresponding end of the field.
+     * For example, `-1` would seek to the last node and `1` would seek to the first.
+     *
+     * NOT allowed if mode is `Fields` and also `pending`.
+     *
+     * TODO: consider splitting out a `firstNode` method, or using a foreEachNode method instead.
+     */
+    seek(offset: number): boolean;
 
-/**
- * A stateful low-level interface for reading tree data.
- *
- * TODO: Needs rules around invalidation/mutation of the underlying tree.
- * Should either be documented here, or each producer should document them
- * (and likely via returning a sub-interface with documentation on the subject).
- *
- * TODO: Needs a way to efficiently clone Cursor so patterns like lazy tree reification can be implemented efficiently.
- *
- * TODO: add optional fast path APIs for more efficient handling when supported by underlying format and reader.
- * Leverage "chunks" and "shape" for this, and skip to next chunk with seek (chunk length).
- * Default chunks of size 1, and "node" shape?
- */
-export interface ITreeCursor<TResult = TreeNavigationResult> {
-    /** Select the child located at the given key and index. */
-    down(key: FieldKey, index: number): TResult;
+    // ********** APIs for when mode = Fields, and not pending ********** //
 
     /**
-     * Moves `offset` entries in the field.
+     * Returns the FieldKey for the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
      */
-    seek(offset: number): TResult;
+    getCurrentFieldKey(): FieldKey;
 
-    /** Select the parent of the currently selected node. */
-    up(): TResult;
+    /**
+     * @returns the number of immediate children in the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     */
+    getCurrentFieldLength(): number;
 
-    /** The type of the currently selected node. */
-    readonly type: TreeType;
+    /**
+     * Sets current node to the node at the provided `index` of the current field.
+     *
+     * Allowed when `mode` is `Fields`, and not `pending`.
+     * Sets mode to `Nodes`.
+     */
+    enterChildNode(index: number): void;
+
+    // ********** APIs for when mode = Nodes ********** //
 
     /**
-     * @returns the keys of the currently selected node.
-     * TODO: ordering invariants: Consistent over time? Consistent across nodes? Sorted?
-     * TODO: empty fields: are they always omitted here? Sometimes omitted? Depends on field kind and schema?
-     * */
-    keys: Iterable<FieldKey>;
+     * @returns a path to the current node.
+     *
+     * Only valid when `mode` is `Nodes`.
+     * Assumes root node is special node above detached sequences.
```

Comment:
This may confuse some people because the "root" node being referred to here is not the same root node that most people think of when they picture a document tree. 
How about:
```suggestion
     * All detached sequences are modeled as being under some imaginary node whose path is undefined.
```
Or something similar?

Reply:
While you comment makes more sense than mine, it says the wrong thing.

This function is making the assumption that the root of the tree this cursor was made from is the "imaginary node whose path is undefined".

This means if you take a free (unrooted) tree and make a cursor out of it (ex: a TextCusror from a random object thats not part of a forest, ex: a tree you could insert into a forest), and call this function, the very much real object you passed to the TextCursor constructor is treated is the "imaginary node whose path is undefined", its fields get wrongly treated as detached sequences, and the path you get back is malformed/nonsense since the documentation on UpPath does not accurately reflect what it does at the root.


My comment is structured as it is because this function makes an assumption about how this cursor is being used, and it is invalid (and thus undefined behavior) to call this function when the assumption is not true. We need to express this usage constraint, not just clarify how roots work.

How about:

```suggestion
     * Assumes this cursor has a root node where its field keys are actually detached sequences.
     * If the cursor is not rooted at such a node, calling this function is invalid, and the returned UpPath (if any) may not be meaningful.
     * This requirement exists because {@link UpPath}s are absolute paths and thus must be rooted in a detached sequence.
     * TODO: consider adding an optional base path to append to remove/clarify this restriction.
```

Reply:
My bad, I forgot we have to deal with cursors to free trees.
Your suggested comment looks good.